### DISCUSSION TITLE
Clerk Auth with Supabase in Nextjs 14

### DISCUSSION THREAD

**Discussion by Pratham3032002** [supabase/supabase#22165] [Questions]
I am building a project in Next Js 14. My current auth system uses supabase's auth functionalities. Everything from session management to db is powered by a supabase. 

I am however trying to switch to clerk for authentication. The problem I am facing is how to link clerk's userID with my supabase auth table. 

I do not want to change any code for session management etc, hence I am trying to figure how to pass user details from clerk to my supabase auth db so the rest of the application can work well.

**Comment by GaryAustin1** [comment]
Not sure this can be done.  Supabase sessions depend on the UUID sub in the access token and the refresh token to work.
Clerk I believe uses a string for the sub claim and the jwt cannot be refreshed by Supabase auth as it is not in the tables involved with keeping track of access and refresh tokens.

You can use createClient with an authorization header set to the Clerk jwt and have RLS work except for auth.uid() function has to be changed ( and any UUID user id columns have to be changed to use the Clerk sting user id.

**Comment by solvionsolutions** [comment]
Since you're building off Gary's answer here, the cleanest path in 2025 is Clerk's native third-party auth integration rather than the old JWT-template route — Clerk deprecated the `supabase` JWT template back in April 2025, so any tutorial telling you to mint one is out of date. With the native integration, Clerk's session token is handed to Supabase directly and the `sub` claim carries the Clerk user id (e.g. `user_2abc...`), which is exactly why your current flow chokes: those ids aren't UUIDs, so a `uuid` column or an `auth.uid()` cast will throw.

On the database side, store your user-id columns as `text` instead of `uuid` and have your RLS policies read the id straight off the JWT:

```sql
create policy "Users manage own rows"
on public.your_table
for all
using  (auth.jwt()->>'sub' = user_id)
with check (auth.jwt()->>'sub' = user_id);
```

That lets the rest of your app stay mostly as-is — you stop trying to fit Clerk into Supabase's auth session and just let RLS trust Clerk's token. And since you're on Next.js 14, you can drop the Supabase session-refresh juggling entirely and pass the Clerk token when you construct the client. It's a bit of upfront migration on the column types, but it's stable once it's in.

I wrote up the full walkthrough here: https://www.guardlayer.io/blog/clerk-supabase-rls
